### PR TITLE
[NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54742 test_c10d: use with_nccl_blocking_wait decorator
* #54741 test_c10d: Run tests with nccl_async_error_handling
* #54740 Provide a decorator to set/unset nccl blocking wait for tests
* **#54558 [NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier
blocking wait**
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* #53773 Gloo-only CPU-based monitored barrier

blocking wait

In blocking wait's polling synchronization loop, we frequently call checkAndSetException() as part of isCompleted() to check the status of nccl operations. It would be useful to log here in case we encounter any exceptions (which are later thrown by `checkAndThrowException`).

Also slightly refactors code previously added to make use of a helper function to get the error message given an `std::exception_ptr`.

Differential Revision: [D27136202](https://our.internmc.facebook.com/intern/diff/D27136202/)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -167,6 +167,18 @@ std::string getNcclAbortedCommStoreKey(const std::string ncclIdStr) {
...
+std::string getExceptionMsgFromExceptionPtr(const std::exception_ptr exceptionPtr) {
+  TORCH_CHECK(exceptionPtr != nullptr);
+  std::string msg;
+  try {
+    std::rethrow_exception(exceptionPtr);
+  } catch (const std::exception &e) {
+    msg = e.what();
+  }
+  return msg;
+}
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:180]: I think you can rewrite this as follows:

```
    try {
      std::rethrow_exception(eptr);
    } catch (const std::exception& e) {
      return e.what();
    } catch (...) {
      return "Unknown Exception Type";
    }
```

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -240,14 +252,26 @@ bool ProcessGroupNCCL::WorkNCCL::isSuccess() const {
...
 
 void ProcessGroupNCCL::WorkNCCL::checkAndSetException() {
-  if (exception()) {
+  auto existingException = exception();
+  if (existingException) {
+    auto existingExceptionMsg = getExceptionMsgFromExceptionPtr(existingException);
+    LOG(INFO) << "[Rank " << rank_ << "]"
+              << " already had existing exception: " << existingExceptionMsg
+              << " when checking for NCCL communicator errors.";
     // We already have an exception.
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:261]: Do we need to LOG here again for the same rank? We would already have a LOG when we set the exception for the first time below right?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -240,14 +252,26 @@ bool ProcessGroupNCCL::WorkNCCL::isSuccess() const {
...
   }
 
   auto exception_ptr = checkForNCCLErrors(ncclComms_);
   std::unique_lock<std::mutex> lock(mutex_);
   exception_ = exception_ptr;
+  if (exception_) {
+    auto msg = getExceptionMsgFromExceptionPtr(exception_);
+    LOG(INFO) << "[Rank " << rank_ << "]"
+              << " found async exception when checking for NCCL errors: "
+              << msg;
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:272]: nit: can just call `getExceptionMsgFromExceptionPtr(exception_);` inline here?

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:261]: That's a good point, we've already logged this one and there's no reason to log it again.

**Review by pritamdamania87 — APPROVED** [review]

**Pull request closed without merging**